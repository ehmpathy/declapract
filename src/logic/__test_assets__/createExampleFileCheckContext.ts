import { FileCheckContext } from '../../domain/objects/FileCheckContext';

export const createExampleFileCheckContext = () =>
  new FileCheckContext({
    projectRootDirectory: '/some/project/root',
    relativeFilePath: 'some/file/path.ts',
    projectVariables: {},
    declaredFileContents: '__DECLARED_CONTENTS__',
  });
