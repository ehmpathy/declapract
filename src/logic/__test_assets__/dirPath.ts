export const testAssetsDirectoryPath = __dirname;
