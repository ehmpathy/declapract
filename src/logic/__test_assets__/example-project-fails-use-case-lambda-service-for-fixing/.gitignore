# all contents of dir are ignored since tests provision and fix the contents in here
*
!.gitignore
!declapract.use.yml

