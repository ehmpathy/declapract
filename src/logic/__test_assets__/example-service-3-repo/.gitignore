.declapract
node_modules
package-lock.json
