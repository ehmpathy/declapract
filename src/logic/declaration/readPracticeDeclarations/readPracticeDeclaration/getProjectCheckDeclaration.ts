import { FileCheckPurpose, FileCheckDeclaration } from '../../../../domain';
import { ProjectCheckDeclaration } from '../../../../domain/objects/ProjectCheckDeclaration';
import { readFileAsync } from '../../../../utils/fileio/readFileAsync';
import { listFilesInDirectory } from '../../../../utils/filepaths/listFilesInDirectory';
import { UnexpectedCodePathError } from '../../../UnexpectedCodePathError';
import { getFileCheckDeclaration } from './getFileCheckDeclaration/getFileCheckDeclaration';

export const getProjectCheckDeclaration = async ({
  purpose,
  declaredProjectDirectory,
}: {
  purpose: FileCheckPurpose;
  declaredProjectDirectory: string;
}): Promise<ProjectCheckDeclaration> => {
  // grab name from the directory
  const name = (() => {
    if (purpose === FileCheckPurpose.BEST_PRACTICE)
      return (new RegExp(/\/([\w-]+)\/best-practice$/).exec(declaredProjectDirectory) ?? [])[1] ?? null;
    if (purpose === FileCheckPurpose.BAD_PRACTICE)
      return (new RegExp(/bad-practices\/([\w-]+)$/).exec(declaredProjectDirectory) ?? [])[1] ?? null;
  })();
  if (!name)
    throw new UnexpectedCodePathError(
      `neither best-practice name nor bad-practice name was extractable from the declared project directory '${declaredProjectDirectory}'`,
    );

  // grab paths to _all_ files in this dir (not just at root level)
  const paths = await listFilesInDirectory({ directory: declaredProjectDirectory });

  // grab the meta files (i.e., path matches `${projectRoot}/.declapract.*`)
  const metaFilePaths = paths.filter((path) => new RegExp(/^\.declapract\./).test(path));

  // group all of the other files by main file name (i.e., key = filePath.replace('.declapract.ts$', ''))
  const projectFilePaths = [
    ...new Set(
      paths.filter((path) => !metaFilePaths.includes(path)).map((path) => path.replace(/\.declapract\.ts$/, '')),
    ),
  ];

  // for each "main file", get the FileCheckDefinition, now that we have all the files defined for it
  const checksAndErrors = await Promise.all(
    projectFilePaths.map((declaredFileCorePath) =>
      getFileCheckDeclaration({ purpose, declaredProjectDirectory, declaredFileCorePath }).catch((error) => error),
    ),
  );
  const anError = checksAndErrors.find((checkOrError) => checkOrError instanceof Error);
  if (anError) throw anError;
  const checks: FileCheckDeclaration[] = checksAndErrors.filter((checkOrError) => !(checkOrError instanceof Error));
  if (!checks.length) {
    console.log({ anError });
    console.log({
      declaredProjectDirectory,
      projectFilePaths,
      checksAndErrorsLength: checksAndErrors.length,
      checksLength: checks.length,
    });
  }

  // get readme contents, if readme defined
  const readme = metaFilePaths.includes('.declapract.readme.md')
    ? await readFileAsync({ filePath: `${declaredProjectDirectory}/.declapract.readme.md` })
    : null;

  // return the project def
  return new ProjectCheckDeclaration({
    name,
    checks,
    readme,
  });
};
