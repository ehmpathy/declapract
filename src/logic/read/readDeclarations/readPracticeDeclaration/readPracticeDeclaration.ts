import { CheckProjectDeclaration, PracticeDeclaration } from '../../../../domain';
import { listPathsInDirectory } from '../../../../utils/filepaths/listPathsInDirectory';
import { UserInputError } from '../../../UserInputError';
import { getCheckProjectDeclaration } from './getCheckProjectDeclaration';

export const readPracticeDeclaration = async ({ declaredPracticeDirectory }: { declaredPracticeDirectory: string }) => {
  // grab the file paths in this dir
  const paths = await listPathsInDirectory(declaredPracticeDirectory);

  // grab the name of the practice
  const practiceName = declaredPracticeDirectory.split('/').slice(-1)[0];

  // find the "best-practice" dir, if one was defined
  const bestPracticeProjectCheck: CheckProjectDeclaration | null = paths.includes('best-practice')
    ? await getCheckProjectDeclaration({ declaredProjectDirectory: `${declaredPracticeDirectory}/best-practice` })
    : null;

  // find the "bad-practice" dirs, if any were defined
  const badPracticeProjectChecks: CheckProjectDeclaration[] = await (async () => {
    if (!paths.includes('bad-practices')) return [];
    const badPracticesPaths = await listPathsInDirectory(`${declaredPracticeDirectory}/bad-practices`);
    return Promise.all(
      badPracticesPaths.map((path) =>
        getCheckProjectDeclaration({ declaredProjectDirectory: `${declaredPracticeDirectory}/bad-practices/${path}` }),
      ),
    );
  })();

  // ensure atleast one of best-practice or bad-practice is defined
  if (!bestPracticeProjectCheck && !badPracticeProjectChecks.length)
    throw new UserInputError(
      `either a 'best-practice' or atleast one of 'bad-practices' must be defined for '${practiceName}'`,
    );

  // return it
  return new PracticeDeclaration({
    name: practiceName,
    bestPractice: bestPracticeProjectCheck,
    badPractices: badPracticeProjectChecks,
  });
};
