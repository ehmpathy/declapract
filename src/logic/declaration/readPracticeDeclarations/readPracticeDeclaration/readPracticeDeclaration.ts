import { ProjectCheckDeclaration, PracticeDeclaration, FileCheckPurpose } from '../../../../domain';
import { listPathsInDirectory } from '../../../../utils/filepaths/listPathsInDirectory';
import { UserInputError } from '../../../UserInputError';
import { getProjectCheckDeclaration } from './getProjectCheckDeclaration';

export const readPracticeDeclaration = async ({ declaredPracticeDirectory }: { declaredPracticeDirectory: string }) => {
  // grab the file paths in this dir
  const paths = await listPathsInDirectory({ directory: declaredPracticeDirectory });

  // grab the name of the practice
  const practiceName = declaredPracticeDirectory.split('/').slice(-1)[0];

  // find the "best-practice" dir, if one was defined
  const bestPracticeProjectCheck: ProjectCheckDeclaration | null = paths.includes('best-practice')
    ? await getProjectCheckDeclaration({
        purpose: FileCheckPurpose.BEST_PRACTICE,
        declaredProjectDirectory: `${declaredPracticeDirectory}/best-practice`,
      })
    : null;

  // find the "bad-practice" dirs, if any were defined
  const badPracticeProjectChecks: ProjectCheckDeclaration[] = await (async () => {
    if (!paths.includes('bad-practices')) return [];
    const badPracticesPaths = await listPathsInDirectory({ directory: `${declaredPracticeDirectory}/bad-practices` });
    return Promise.all(
      badPracticesPaths.map((path) =>
        getProjectCheckDeclaration({
          purpose: FileCheckPurpose.BAD_PRACTICE,
          declaredProjectDirectory: `${declaredPracticeDirectory}/bad-practices/${path}`,
        }),
      ),
    );
  })();

  // make sure that the `/bad-practices` dir was not typo'ed as `/bad-practice` on accident (happened on more than one occasion ðŸ˜…)
  if (paths.includes('bad-practice'))
    throw new UserInputError(
      `a 'bad-practice/' directory was defined for practice '${practiceName}'. this is probably a typo and should have been 'bad-practices/'`,
      { potentialSolution: "rename 'bad-practice/' to 'bad-practices/'" },
    );

  // ensure atleast one of best-practice or bad-practice is defined
  if (!bestPracticeProjectCheck && !badPracticeProjectChecks.length)
    throw new UserInputError(
      `either a 'best-practice' or atleast one of 'bad-practices' must be defined for '${practiceName}'`,
    );

  // return it
  return new PracticeDeclaration({
    name: practiceName,
    bestPractice: bestPracticeProjectCheck,
    badPractices: badPracticeProjectChecks,
  });
};
