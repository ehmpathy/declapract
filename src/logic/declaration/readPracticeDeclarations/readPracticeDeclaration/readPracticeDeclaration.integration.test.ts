import { testAssetsDirectoryPath } from '../../../__test_assets__/dirPath';
import { readPracticeDeclaration } from './readPracticeDeclaration';

describe('readPracticeDeclarationFromDirectory', () => {
  it('should define correctly when only the best-practice is defined', async () => {
    const practice = await readPracticeDeclaration({
      declaredPracticeDirectory: `${testAssetsDirectoryPath}/example-best-practices-repo/src/practices/util-sleep`,
    });
    // console.log(JSON.stringify(practice, null, 2));
    expect(practice.bestPractice).not.toEqual(null);
    expect(practice.badPractices.length).toEqual(0);
    expect(practice).toMatchSnapshot();
  });
  it('should define correctly when a bad-practice is defined', async () => {
    const practice = await readPracticeDeclaration({
      declaredPracticeDirectory: `${testAssetsDirectoryPath}/example-best-practices-repo/src/practices/dates-and-times`,
    });
    expect(practice.bestPractice).not.toEqual(null);
    expect(practice.badPractices.length).toEqual(1);
    expect(practice).toMatchSnapshot();
  });
  it('should define correctly when best and bad practices are defined', async () => {
    const practice = await readPracticeDeclaration({
      declaredPracticeDirectory: `${testAssetsDirectoryPath}/example-best-practices-repo/src/practices/directory-structure-src`,
    });
    expect(practice.bestPractice).not.toEqual(null);
    expect(practice.badPractices.length).toEqual(2);
    expect(practice).toMatchSnapshot();
  });
  it('should still have the fix functions defined on the file checks', async () => {
    const practice = await readPracticeDeclaration({
      declaredPracticeDirectory: `${testAssetsDirectoryPath}/example-best-practices-repo/src/practices/prettier`,
    });
    expect(practice.bestPractice?.checks[0].fix).toBeDefined();
    expect(practice.bestPractice?.checks[0].fix).not.toEqual(null);
    expect(practice.bestPractice).not.toEqual(null);
    expect(practice.badPractices.length).toEqual(0);
    expect(practice).toMatchSnapshot();
  });
});
