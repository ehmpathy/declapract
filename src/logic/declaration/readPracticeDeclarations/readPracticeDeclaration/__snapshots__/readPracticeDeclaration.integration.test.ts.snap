// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readPracticeDeclarationFromDirectory should define correctly when a bad-practice is defined 1`] = `
PracticeDeclaration {
  "badPractices": Array [
    ProjectCheckDeclaration {
      "checks": Array [
        FileCheckDeclaration {
          "check": [Function],
          "contents": null,
          "fix": null,
          "pathGlob": "package.json",
          "purpose": "BAD_PRACTICE",
          "required": true,
          "type": "CUSTOM",
        },
      ],
      "name": "moment",
      "readme": "using \`moment\` is a bad practice because it has a weird api that \\"mutates\\" values

for example:
\`\`\`ts
const a = moment();
const b = a.addHours(7);
expect(b.toISOString()).not.toEqual(a.toISOString()); // this will fail, since moment actually both returns \`b\` _and_ mutates \`a\`...
\`\`\`

",
    },
  ],
  "bestPractice": ProjectCheckDeclaration {
    "checks": Array [
      FileCheckDeclaration {
        "check": [Function],
        "contents": null,
        "fix": null,
        "pathGlob": "package.json",
        "purpose": "BEST_PRACTICE",
        "required": true,
        "type": "CUSTOM",
      },
    ],
    "name": "dates-and-times",
    "readme": "\`date-fns\` is a great library for manipulating dates and times in a functional, immutable, declarative pattern
",
  },
  "name": "dates-and-times",
}
`;

exports[`readPracticeDeclarationFromDirectory should define correctly when best and bad practices are defined 1`] = `
PracticeDeclaration {
  "badPractices": Array [
    ProjectCheckDeclaration {
      "checks": Array [
        FileCheckDeclaration {
          "check": [Function],
          "contents": null,
          "fix": [Function],
          "pathGlob": "src/model/**/*.ts",
          "purpose": "BAD_PRACTICE",
          "required": true,
          "type": "EXISTS",
        },
      ],
      "name": "model-dir",
      "readme": "we use \`domain\` directory instead of \`model\` directory
- model directory was what we originally went with, but realized \`domain\` was more explicit and better represented what we were going after
",
    },
    ProjectCheckDeclaration {
      "checks": Array [
        FileCheckDeclaration {
          "check": [Function],
          "contents": null,
          "fix": [Function],
          "pathGlob": "src/services/**/*.ts",
          "purpose": "BAD_PRACTICE",
          "required": true,
          "type": "EXISTS",
        },
      ],
      "name": "services-dir",
      "readme": null,
    },
  ],
  "bestPractice": ProjectCheckDeclaration {
    "checks": Array [
      FileCheckDeclaration {
        "check": [Function],
        "contents": null,
        "fix": null,
        "pathGlob": "src/contract/**/*.ts",
        "purpose": "BEST_PRACTICE",
        "required": true,
        "type": "EXISTS",
      },
      FileCheckDeclaration {
        "check": [Function],
        "contents": "import { invokeLambdaFunction } from 'simple-lambda-client';
",
        "fix": [Function],
        "pathGlob": "src/data/clients/**/*.ts",
        "purpose": "BEST_PRACTICE",
        "required": false,
        "type": "CONTAINS",
      },
      FileCheckDeclaration {
        "check": [Function],
        "contents": null,
        "fix": null,
        "pathGlob": "src/data/dao/**/*.ts",
        "purpose": "BEST_PRACTICE",
        "required": false,
        "type": "EXISTS",
      },
      FileCheckDeclaration {
        "check": [Function],
        "contents": null,
        "fix": null,
        "pathGlob": "src/domain/constants.ts",
        "purpose": "BEST_PRACTICE",
        "required": false,
        "type": "EXISTS",
      },
      FileCheckDeclaration {
        "check": [Function],
        "contents": null,
        "fix": null,
        "pathGlob": "src/domain/index.ts",
        "purpose": "BEST_PRACTICE",
        "required": true,
        "type": "EXISTS",
      },
      FileCheckDeclaration {
        "check": [Function],
        "contents": null,
        "fix": null,
        "pathGlob": "src/domain/objects/index.ts",
        "purpose": "BEST_PRACTICE",
        "required": true,
        "type": "EXISTS",
      },
      FileCheckDeclaration {
        "check": [Function],
        "contents": null,
        "fix": null,
        "pathGlob": "src/logic/**/*.ts",
        "purpose": "BEST_PRACTICE",
        "required": true,
        "type": "EXISTS",
      },
    ],
    "name": "directory-structure-src",
    "readme": "layers to a project:
- contract layer
  - summary: responsible for strictly exposing the logic/domain defined in the service
  - responsibilities:
    - specifies the inputs and outputs
    - normalizes inputs on behalf of user
    - validates the inputs
    - normalizes outputs on behalf of service
- domain layer
  - summary: defines the domain which this service masters
  - responsibilities:
    - specify domain objects
    - specify constants
- logic layer
  - summary: defines the business logic that this service manages
  - responsibilities:
    - define the business logic of the domain in an organized and maintainable way
- data layer:
  - summary: defines how to persist data / leverage third party clients
  - responsibilities:
    - daos
    - clients
",
  },
  "name": "directory-structure-src",
}
`;

exports[`readPracticeDeclarationFromDirectory should define correctly when only the best-practice is defined 1`] = `
PracticeDeclaration {
  "badPractices": Array [],
  "bestPractice": ProjectCheckDeclaration {
    "checks": Array [
      FileCheckDeclaration {
        "check": [Function],
        "contents": "export const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));
",
        "fix": [Function],
        "pathGlob": "src/**/sleep.ts",
        "purpose": "BEST_PRACTICE",
        "required": false,
        "type": "EQUALS",
      },
    ],
    "name": "util-sleep",
    "readme": "checks that if a file named \`sleep.ts\` exists, then it must match the proper definition for our sleep utility, defined as a best practice

optional check => if you dont need \`sleep.ts\`, dont define the file

helpful in that:
- it makes sure your definition of sleep is correct
- you can just define an empty file called \`sleep.ts\` and then run \`declapract fix --practice:util-sleep\` and it will fill it out for you
",
  },
  "name": "util-sleep",
}
`;

exports[`readPracticeDeclarationFromDirectory should still have the fix functions defined on the file checks 1`] = `
PracticeDeclaration {
  "badPractices": Array [],
  "bestPractice": ProjectCheckDeclaration {
    "checks": Array [
      FileCheckDeclaration {
        "check": [Function],
        "contents": "*.md
",
        "fix": [Function],
        "pathGlob": ".prettierignore",
        "purpose": "BEST_PRACTICE",
        "required": true,
        "type": "EQUALS",
      },
      FileCheckDeclaration {
        "check": [Function],
        "contents": "{
  \\"devDependencies\\": {
    \\"prettier\\": \\"@declapract{check.minVersion('2.0.0')}\\"
  },
  \\"scripts\\": {
    \\"format\\": \\"prettier --write '**/*.ts' --config ./prettier.config.js\\"
  }
}
",
        "fix": [Function],
        "pathGlob": "package.json",
        "purpose": "BEST_PRACTICE",
        "required": true,
        "type": "CONTAINS",
      },
      FileCheckDeclaration {
        "check": [Function],
        "contents": "// ref: http://json.schemastore.org/prettierrc

module.exports = {
  trailingComma: 'all',
  tabWidth: 2,
  singleQuote: true,
  printWidth: 150,
};
",
        "fix": [Function],
        "pathGlob": "prettier.config.js",
        "purpose": "BEST_PRACTICE",
        "required": true,
        "type": "EQUALS",
      },
    ],
    "name": "prettier",
    "readme": null,
  },
  "name": "prettier",
}
`;
